demos/common/http_demo_helpers/http_demo_utils.c

```diff
@@ -26,27 +26,129 @@
 #include "http_demo_utils.h"
 
 /* Retry utilities. */
-#include "retry_utils.h"
+#include "backoff_algorithm.h"
+
+/* Include PKCS11 helpers header. */
+#include "pkcs11_helpers.h"
 
 /* Parser utilities. */
 #include "http_parser.h"
 
 /*-----------------------------------------------------------*/
 
+/**
+ * @brief The maximum number of retries for connecting to server.
+ */
+#define CONNECTION_RETRY_MAX_ATTEMPTS            ( 5U )
+
+/**
+ * @brief The maximum back-off delay (in milliseconds) for retrying connection to server.
+ */
+#define CONNECTION_RETRY_MAX_BACKOFF_DELAY_MS    ( 5000U )
+
+/**
+ * @brief The base back-off delay (in milliseconds) to use for connection retry attempts.
+ */
+#define CONNECTION_RETRY_BACKOFF_BASE_MS         ( 500U )
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief A global storage variable that is used by the pseudo random number generator.
+ * The random number generator is used for calculating exponential back-off
+ * with jitter for retry attempts of failed network operations with the broker.
+ */
+static uint32_t ulNextRand;
+
+/**
+ * @brief A global flag that represents whether the random number
+ * generator has been seeded.
+ */
+static BaseType_t xRngSeeded = pdFALSE;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The random number generator to use for exponential backoff with
+ * jitter retry logic.
+ * This function is an implementation the #BackoffAlgorithm_RNG_t interface type
+ * of the backoff algorithm library API.
+ *
+ * @return The generated random number. This function ALWAYS succeeds
+ * in generating a random number.
+ */
+static int32_t prvGenerateRandomNumber();
+
+/**
+ * @brief Seed the random number generator used for exponential backoff
+ * used in connection retry attempts.
+ *
+ * The PKCS11 module is used to seed the random number generator
+ * as it allows access to a True Random Number Generator (TRNG) if the
+ * vendor platform supports it.
+ * It is recommended to use a device-specific unique entropy source
+ * to seed the random number generator so that probability of collisions
+ * from devices in connection retries is mitigated.
+ *
+ * @return pdPASS if seeding the randon number generator was successful;
+ * otherwise pdFAIL for failure.
+ */
+static BaseType_t prvSeedRandomNumberGenerator();
+
+/*-----------------------------------------------------------*/
+
+static int32_t prvGenerateRandomNumber()
+{
+    const uint32_t ulMultiplier = 0x015a4e35UL, ulIncrement = 1UL;
```

Comment:
Why does this routine do anything other than call `xPkcs11GenerateRandomNumber()`?  The RNG code on lines 104-105 looks to have issues (see comment below).  If you just call the PKCS#11 routine you can delete that.

Reply:
Have answered the reason of not using PKCS11 in every iteration of random number generation in the reply [here](https://github.com/aws/amazon-freertos/pull/2767#discussion_r529068149)